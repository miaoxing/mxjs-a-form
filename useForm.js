import React, {useContext} from 'react';
import FormContext from './FormContext';

export default function useForm() {
  return useContext(FormContext);
}
