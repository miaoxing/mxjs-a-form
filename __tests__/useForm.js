import {render} from '@testing-library/react';
import {MemoryRouter} from 'react-router';
import {Form, FormItem} from '..';
import { createRef, useEffect } from 'react';
import $ from 'miaoxing';
import {createPromise} from '@mxjs/test';
import useForm from '../useForm';

const Custom = () => {
  const form = useForm();

  useEffect(() => {
    form.setInputConverter(values => {
      values.foo = 'abc';
      return values;
    });
    form.setOutputConverter(values => {
      delete values.foo;
      values.bar = 'bbc';
      return values;
    });
  }, []);

  return '';
};

describe('useForm', () => {
  test('converter', async () => {
    const promise = createPromise();
    const promise2 = createPromise();
    $.http = jest.fn().mockImplementationOnce(() => promise.resolve({
      ret: {
        code: 1,
        message: 'success',
        data: {},
      },
    })).mockImplementation(() => promise2.resolve({
      ret: {
        code: 1,
        message: 'success',
        data: {},
      },
    }));

    const form = createRef();
    const {container, findByDisplayValue} = render(<MemoryRouter>
      <Form
        formRef={form}
      >
        <Custom/>
        <FormItem name="foo"/>
      </Form>
    </MemoryRouter>);

    await findByDisplayValue('abc');
    expect(container.querySelector('#foo').value).toBe('abc');

    form.current.submit();
    await promise2;

    expect($.http).toMatchSnapshot();
  });
});
