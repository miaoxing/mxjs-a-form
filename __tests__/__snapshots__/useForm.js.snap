// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useForm converter 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "method": "get",
        "url": "",
      },
    ],
    Array [
      Object {
        "data": Object {
          "bar": "bbc",
        },
        "loading": true,
        "method": "POST",
        "url": "",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;
